### PR TITLE
filter: use selector to filter the non-regex string

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

Part of https://github.com/pingcap/dm/issues/404

If the given string is not regex. We use selector to filter it so it can support simple pattern like `?`, `*` or `[]` after #302 is merged.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects

 - Increased code complexity

### CODE REVIEW

**Review by amyangfei — COMMENTED** [review]
Please fix the unit test, seems it fails because of the slice value in trie node

**Comment by winoros** [comment]
@amyangfei Fixed.

**Comment by IANTHEREAL** [comment]
@winoros please resolve conflicts

File: pkg/filter/filter.go [hunk 1]
```diff
@@ -107,7 +109,12 @@ type Filter struct {
 
 // New creates a filter use the rules.
 func New(caseSensitive bool, rules *Rules) (*Filter, error) {
+	if !caseSensitive {
```
Comment (IANTHEREAL) [inline pkg/filter/filter.go:112]: It also  does lowercase conversion in `initRules`, can we remove it here?

File: pkg/filter/filter.go [hunk 2]
```diff
@@ -187,6 +207,75 @@ func (f *Filter) addOneRegex(originStr string) error {
 	return nil
 }
 
+func (f *Filter) initSchemaRule(dbStr string, isWhiteList bool) error {
+	if strings.HasPrefix(dbStr, "~") {
+		err := f.initOneRegex(dbStr[1:])
+		if err != nil {
```
Comment (IANTHEREAL) [inline pkg/filter/filter.go]: return err directly if err is nil

**Comment by IANTHEREAL** [comment]
Rest LGTM

File: pkg/filter/filter_test.go [hunk 3]
```diff
@@ -107,10 +102,36 @@ func (s *testFilterSuite) TestFilterOnSchema(c *C) {
...
+		// ensure case-insensitive
+		{
+			rules: &Rules{
+				IgnoreDBs:    []string{"~^FOO"},
+				IgnoreTables: []*Table{{"~.*", "~FoO$"}},
+			},
+			Input:         []*Table{{"FOO1", "a"}, {"foo2", "b"}, {"BoO3", "cFoo"}, {"Foo4", "dfoo"}, {"5", "5"}},
+			Output:        []*Table{{"foo2", "b"}, {"BoO3", "cFoo"}, {"Foo4", "dfoo"}, {"5", "5"}},
+			caseSensitive: true,
+		},
```
Comment (amyangfei) [inline pkg/filter/filter_test.go:114]: what about adding more tests for case-insensitive/non case -insensitive test with wildcard

File: pkg/filter/filter_test.go [hunk 4]
```diff
@@ -107,10 +102,36 @@ func (s *testFilterSuite) TestFilterOnSchema(c *C) {
...
+		// test the rule whose schema part is not regex and the table part is regex.
+		{
+			rules: &Rules{
+				IgnoreTables: []*Table{{"a?b?", "~f[0-9]"}},
+			},
+			Input:  []*Table{{"abbd", "f1"}, {"aaaa", "f2"}, {"5", "5"}, {"abbc", "fa"}},
+			Output: []*Table{{"aaaa", "f2"}, {"5", "5"}, {"abbc", "fa"}},
+		},
+		// test the rule whose schema part is regex abd the table part is not regex.
```
Comment (amyangfei) [inline pkg/filter/filter_test.go]:
```suggestion
		// test the rule whose schema part is regex and the table part is not regex.
```

File: pkg/filter/filter.go [hunk 5]
```diff
@@ -244,41 +337,69 @@ func (f *Filter) filterOnTables(tb *Table) bool {
...
-func (f *Filter) findMatchedDoTables(tb *Table) bool {
-	return f.matchTable(f.rules.DoTables, tb)
-}
-
-func (f *Filter) findMatchedIgnoreTables(tb *Table) bool {
-	return f.matchTable(f.rules.IgnoreTables, tb)
-}
-
-func (f *Filter) matchDB(patternDBS []string, a string) bool {
+func (f *Filter) matchDB(patternDBS []string, a string, isWhileListCheck bool) bool {
```
Comment (amyangfei) [inline pkg/filter/filter.go]:
```suggestion
func (f *Filter) matchDB(patternDBS []string, a string, isWhiteListCheck bool) bool {
```

**Comment by amyangfei** [comment]
friendly ping @winoros

**Review by amyangfei — COMMENTED** [review]
LGTM

**Comment by amyangfei** [comment]
PTAL @GregoryIan

**Comment by IANTHEREAL** [comment]
please fix conflicts @winoros

**Comment by amyangfei** [comment]
Friendly ping, please resolve conflicts @winoros

**Comment by IANTHEREAL** [comment]
let me take over this PR 🤷‍♂

**Comment by IANTHEREAL** [comment]
LGTM

**Comment by IANTHEREAL** [comment]
@amyangfei  PTAL again, and approve it if there's not any other problems

File: pkg/filter/filter.go [hunk 6]
```diff
@@ -198,25 +277,25 @@ func (f *Filter) ApplyOn(stbs []*Table) []*Table {
...
+		if do {
 			tbs = append(tbs, tb)
 		}
 	}
 
 	return tbs
 }
 
-// Match returns true if the specified table should not be removed.
-func (f *Filter) Match(tb *Table) bool {
```
Comment (amyangfei) [inline pkg/filter/filter.go]: This function can be kept, and ditto for the test case

**Review by amyangfei — COMMENTED** [review]
Rest LGTM

**Review by amyangfei — APPROVED** [review]
LGTM

**Pull request merged**